hal/src/dma.rs

```diff
@@ -0,0 +1,1075 @@
+        // self.job_status.set(Status::DmaStatusOk);
+        // let channel = self.channel.unwrap();
+        //
+        // #[cfg(feature = "samd51")]
+        // DMAC.chctrla(channel).bit.ENABLE = 1;
+        //
+        // #[cfg(not(feature = "samd51"))]
+        // {
+        //    CHID.bit.ID = channel;
+        //    CHCTRLA.reg |= CHCTRLA_ENABLE;
+        // }
+        //
+        // cpu_irq_leave_critical();
+    }
+
+    // TODO(adafruit): delete descriptor, delete whole descriptor chain
+
+    /// Select whether channel's descriptor list should repeat or not.
+    /// This can be done before or after channel & any descriptors are allocated.
+    pub fn enable_loop(&mut self, flag: bool) {
+        // The loop selection is 'sticky' -- that is, you can enable or
+        // disable looping before a descriptor list is built, or after
+        // the fact.  This requires some extra steps in the library code
+        // but avoids a must-do-in-X-order constraint on user.
+        self.loop_flag = flag;
+
+        if self.has_descriptors {
+            // Descriptor list already started?
+            let channel = self.channel.expect("no channel ID");
+
+            // Scan descriptor list to find last entry.  If an entry's
+            // DESCADDR value is 0, that's the end of the list and it's
+            // currently un-looped.  If the DESCADDR value is the same
+            // as the first entry, that's the end of the list and it's
+            // already looped.
+            let mut desc = self.descriptor_list.descriptor(channel);
+
+            loop {
+                let next_desc = desc.descaddr.get();
+
+                if next_desc != 0
+                    && next_desc
+                        != (self.descriptor_list.descriptor(channel) as *const Descriptor) as u32
+                {
+                    desc = unsafe { mem::transmute(next_desc) };
+                } else {
+                    break;
+                }
+            }
+
+            // Loop or unloop descriptor list as appropriate
+            if self.loop_flag {
+                desc.descaddr
+                    .set((self.descriptor_list.descriptor(channel) as *const Descriptor) as u32);
+            } else {
+                desc.descaddr.set(0);
+            };
+        }
+    }
+
+    /// Is this DMA currently active?
+    pub fn is_active(&self) -> bool {
+        if let Some(channel) = self.channel {
+            self.descriptor_list.writeback(channel).btctrl.get() & BTCTRL_VALID != 0
+        } else {
+            false
+        }
+    }
+}
+
+/// Thunk for `__get_PRIMASK()`
+fn get_primask() -> u32 {
+    // TODO(tarcieri): implement this!
+    unimplemented!();
+}
+
+/// Thunk for `__enable_irq()`
+fn enable_irq() {
+    // TODO(tarcieri): implement this!
+    unimplemented!();
+}
+
+/// Thunk for `__disable_irq()`
+fn disable_irq() {
+    // TODO(tarcieri): implement this!
+    unimplemented!();
+}
+
+/// Thunk for `__DMB()`
```

Comment:
TODO: `__DMB()` equivalent

Reply:
https://docs.rs/cortex-m/0.2.12/cortex_m/asm/fn.dmb.html

Reply:
Aha, in that case, would it be better to substitute a Rust memory fence instead?

According to Godbolt[1] this program:

```
#![no_std]

use core::sync::atomic;

pub fn test() {
    atomic::fence(atomic::Ordering::SeqCst);
}
```

...generates this thumbv7 assembly:

```
example::test:
        dmb     sy
        bx      lr
```

[1]: https://godbolt.org/z/2SnyjA

Reply:
I get the same ASM for:

```
atomic::fence(atomic::Ordering::Acquire);
```

...which I'm already using elsewhere (cargo culted from the ST DMA driver)

Reply:
Switched to cortex_m::interrupt::Mutex